app/models/settings/general.rb

```diff
@@ -131,6 +131,16 @@ class General < RailsSettings::Base
     # Feed
     field :feed_pinned_article_id, type: :integer, validates: { existing_published_article_id: true, allow_nil: true }
 
+    # SMTP Transactional email service
+    field :smtp_settings, type: :hash, default: {
+      address: ApplicationConfig["SMTP_ADDRESS"],
+      port: ApplicationConfig["SMTP_PORT"],
+      authentication: ApplicationConfig["SMTP_AUTHENTICATION"],
+      user_name: ApplicationConfig["SMTP_USER_NAME"],
+      password: ApplicationConfig["SMTP_PASSWORD"],
+      domain: ApplicationConfig["SMTP_DOMAIN"]
+    }
```

Comment:
I think that keeping them as hash feels "cleaner" but I have no idea what's the performance or developer experience implication. Is there something I'm missing? would we benefit from having them as separate fields?

In terms of where to put them, I didn't consider any of that as I'm rushing this PR through. There are already some "emails" related configuration in place so `Settings::SMTP` would be preferable.

Reply:
@maestromac I don't know enough about how `rails-settings-cached` works to say which approach is better for that specific gem — I've never used it before.

Validations may be a reason to move them out of a hash. I don't know how well validations would work on this hash, but it appears to be trivial on single scalar (string/numeric) values. For example, there doesn't seem to be anything stopping us from saying the port number is `"lol"`, which will cause a runtime error when we try to send an email because we can't connect to `smtp.google.com:lol`.

_However_, if they're all stored separately I'm not sure it would be feasible to validate that all of them are set since validations would be applied to each scalar value individually rather than collectively, like we could if they're colocated in a hash. This doesn't seem like an easy decision.

Reply:
Your call @maestromac

We have a generator that can be used to create new "settings" classes, see https://github.com/forem/forem/blob/main/lib/generators/settings_model/USAGE

The result is a separate class that's also acts like any other model class:

```ruby
> Settings::General.ancestors.include?(ActiveRecord::Base)
=> true
```

that said, you can validate the hash, using `validates: ...`

Here's one example:

https://github.com/forem/forem/blob/df3a9ec3e86ad7c1ab181d03bd10ddf6b2f97f33/app/models/settings/general.rb#L132

it uses the custom validator:

https://github.com/forem/forem/blob/main/app/validators/existing_published_article_id_validator.rb

so you could write a custom `SMTPSettingsValidator` and use it there :)

Reply:
Thank you both for your insights! I went ahead and extracted this out into its own settings model/controller.

Reply:
@jgaskins validation was a great idea! if the user enters a string it's automatically converted to an integer in the db.